include/os/freebsd/spl/sys/uio.h

```diff
@@ -39,31 +39,43 @@
 
 #define	uio_loffset	uio_offset
 
-typedef	struct uio	uio_t;
 typedef	struct iovec	iovec_t;
 typedef	enum uio_seg	uio_seg_t;
 
+typedef struct zfs_uio {
+	struct uio	*uio;
+} uio_t;
+
```

Comment:
Just curious, wouldn't an anonymous union struct work here to wrap OS uio?

Reply:
What did you have in mind? I thought of approaching this with a completely unified uio_t struct; however, there is still separation in FreeBSD vs Linux uio code for OS specific `uio` functions. I figured still keeping them separate worked to allow for a clearer separation in the OS independent parts.

The main driver for doing this was because FreeBSD defines `struct uio` in its headers, and I was wanting to add some member variables to it to allow for the Direct IO work to map user pages in the VFS layers.

Reply:
I think a union could be used, but this seems preferable to me since I'd expect it to allow us to keep any platform specific includes out of the common code.  @lundman do you have a feeling for what would be best for macos?

Reply:
The idea of `zfs_uio_t` is exciting, especially since macOS defines `typedef struct uio *uio_t` and ZFS uses `typedef struct uio uio_t` - so there is a bunch of header hackery to make that work, which would then not be required any more.